crypto/s2n_ecc_evp.c

```diff
@@ -0,0 +1,283 @@
+    S2N_ERROR_IF(ret != out->size, S2N_ERR_ECDHE_SERIALIZING);
+    return 0;
+}
+
+static EC_POINT *s2n_ecc_evp_blob_to_point(struct s2n_blob *blob, const EC_KEY *ec_key) {
+    const EC_GROUP *group = EC_KEY_get0_group(ec_key);
+    EC_POINT *point = EC_POINT_new(group);
+    if (point == NULL) {
+        S2N_ERROR_PTR(S2N_ERR_ECDHE_UNSUPPORTED_CURVE);
+    }
+    if (EC_POINT_oct2point(group, point, blob->data, blob->size, NULL) != 1) {
+        EC_POINT_free(point);
+        S2N_ERROR_PTR(S2N_ERR_BAD_MESSAGE);
+    }
+    return point;
+}
+#endif
+
+int s2n_ecc_evp_read_params_point(struct s2n_stuffer *in, int point_size, struct s2n_blob *point_blob) {
+    notnull_check(in);
+    notnull_check(point_blob);
+    gte_check(point_size, 0);
+
+    /* Extract point from stuffer */
+    point_blob->size = point_size;
+    point_blob->data = (uint8_t *)s2n_stuffer_raw_read(in, point_size);
+    notnull_check(point_blob->data);
+
+    return 0;
+}
+
+int s2n_ecc_evp_write_params_point(struct s2n_ecc_evp_params *ecc_evp_params, struct s2n_stuffer *out) {
+    notnull_check(ecc_evp_params);
+    notnull_check(ecc_evp_params->negotiated_curve);
+    notnull_check(ecc_evp_params->evp_pkey);
+    notnull_check(out);
+
+#if MODERN_EC_SUPPORTED
+    out->blob.size = EVP_PKEY_get1_tls_encodedpoint(ecc_evp_params->evp_pkey, &out->blob.data);
+    S2N_ERROR_IF(out->blob.size != ecc_evp_params->negotiated_curve->share_size, S2N_ERR_ECDHE_SERIALIZING);
+
+    out->blob.data = s2n_stuffer_raw_write(out, out->blob.size);
+    notnull_check(out->blob.data);
+#else
+    uint8_t point_len;
+    DEFER_CLEANUP(EC_KEY *ec_key = EVP_PKEY_get1_EC_KEY(ecc_evp_params->evp_pkey), EC_KEY_free_pointer);
+    S2N_ERROR_IF(ec_key == NULL, S2N_ERR_ECDHE_UNSUPPORTED_CURVE);
+
+    const EC_POINT *point = EC_KEY_get0_public_key(ec_key);
+    const EC_GROUP *group = EC_KEY_get0_group(ec_key);
+    S2N_ERROR_IF(point == NULL || group == NULL, S2N_ERR_ECDHE_UNSUPPORTED_CURVE);
+
+    GUARD(s2n_ecc_evp_calculate_point_length(point, group, &point_len));
+    S2N_ERROR_IF(point_len != ecc_evp_params->negotiated_curve->share_size, S2N_ERR_ECDHE_SERIALIZING);
+    out->blob.data = s2n_stuffer_raw_write(out, point_len);
+    out->blob.size = point_len;
+
+    notnull_check(out->blob.data);
+    GUARD(s2n_ecc_evp_write_point_data_snug(point, group, &out->blob));
+#endif
+    return 0;
+}
+
+int s2n_ecc_evp_parse_params_point(struct s2n_blob *point_blob, struct s2n_ecc_evp_params *ecc_evp_params) {
+    notnull_check(point_blob->data);
+    notnull_check(ecc_evp_params->negotiated_curve);
+    S2N_ERROR_IF(point_blob->size != ecc_evp_params->negotiated_curve->share_size, S2N_ERR_ECDHE_SERIALIZING);
+
+    if (ecc_evp_params->evp_pkey == NULL) {
+        ecc_evp_params->evp_pkey = EVP_PKEY_new();
+    }
+
+    S2N_ERROR_IF(ecc_evp_params->evp_pkey == NULL, S2N_ERR_BAD_MESSAGE);
+
+#if MODERN_EC_SUPPORTED
```

Comment:
There are a lot of preprocessor guards here.  Would it be possible to split out the modern EC stuff without using the macro guards?  Maybe have a `s2n_foo_with_modern_EC()` and `s2n_foo_without_modern_ec`, and just call the right one?

Reply:
We could split it out, but even in that case we would still have to have the `#if MODERN_EC_SUPPORTED` around that function  call as the APIs used are not present in the older versions of OpenSSL (<= OpenSSL 1.0.2), and would fail at compile time.